github.com/cockroachdb/pebble/internal/metamorphic: TestMeta failed

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/17304947?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/17304947?buildTab=artifacts#meta) on refs/heads/master @ [8079611f00bc](https://github.com/cockroachdb/pebble/commits/8079611f00bc):


```
=== CONT  TestMeta/execution/standard-021
=== RUN   TestMeta/execution/standard-027
=== PAUSE TestMeta/execution/standard-027
=== CONT  TestMeta/execution/standard-027
=== RUN   TestMeta/execution/standard-011
=== PAUSE TestMeta/execution/standard-011
=== CONT  TestMeta/execution/standard-011
=== RUN   TestMeta/execution/standard-020
=== PAUSE TestMeta/execution/standard-020
=== CONT  TestMeta/execution/standard-020
=== RUN   TestMeta/compare/standard-002
=== RUN   TestMeta/execution/random-000
=== PAUSE TestMeta/execution/random-000
=== CONT  TestMeta/execution/random-000
=== RUN   TestMeta/execution/random-004
=== PAUSE TestMeta/execution/random-004
=== CONT  TestMeta/execution/random-004
=== RUN   TestMeta/execution/random-006
=== PAUSE TestMeta/execution/random-006
=== CONT  TestMeta/execution/random-006
=== RUN   TestMeta/execution/standard-007
=== PAUSE TestMeta/execution/standard-007
=== CONT  TestMeta/execution/standard-007
=== RUN   TestMeta/compare/standard-013
=== RUN   TestMeta/execution/standard-024
=== PAUSE TestMeta/execution/standard-024
=== CONT  TestMeta/execution/standard-024
=== RUN   TestMeta/execution/standard-026
=== PAUSE TestMeta/execution/standard-026
=== CONT  TestMeta/execution/standard-026
=== RUN   TestMeta/execution/standard-008
=== PAUSE TestMeta/execution/standard-008
=== CONT  TestMeta/execution/standard-008
=== RUN   TestMeta/compare/standard-001
=== RUN   TestMeta/execution/random-009
=== PAUSE TestMeta/execution/random-009
=== CONT  TestMeta/execution/random-009
=== RUN   TestMeta/execution/random-013
=== PAUSE TestMeta/execution/random-013
=== CONT  TestMeta/execution/random-013
=== RUN   TestMeta/execution/random-018
=== PAUSE TestMeta/execution/random-018
=== CONT  TestMeta/execution/random-018
=== RUN   TestMeta/execution/random-022
=== PAUSE TestMeta/execution/random-022
=== CONT  TestMeta/execution/random-022
=== RUN   TestMeta/compare/standard-006
=== RUN   TestMeta/execution/standard-016
=== PAUSE TestMeta/execution/standard-016
=== CONT  TestMeta/execution/standard-016
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMeta$' ./internal/metamorphic -seed 1728879075896495783 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: PEBBLE-277

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/17348835?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/17348835?buildTab=artifacts#meta) on refs/heads/master @ [1ed55247e491](https://github.com/cockroachdb/pebble/commits/1ed55247e491):


```
        // 04:58:03.064 INFO: [JOB 8473] compacting(default) L4 [037071] (989B) Score=5.96 + L5 [029362] (1.0KB) Score=0.67; OverlappingRatio: Single 1.08, Multi 0.00
        // 04:58:03.066 INFO: [JOB 8471] sstable deleted 037061
        // 04:58:03.066 INFO: [JOB 8473] compacting: sstable created 037097
        // 04:58:03.069 INFO: [JOB 8473] compacted(default) L4 [037071] (989B) Score=5.96 + L5 [029362] (1.0KB) Score=0.67 -> L5 [037097] (989B), in 0.0s (0.0s total), output rate 621KB/s
        // 04:58:03.844 INFO: [JOB 8475] compacting(default) L4 [037074] (989B) Score=5.86 + L5 [029331] (1.0KB) Score=0.67; OverlappingRatio: Single 1.08, Multi 0.00
        // 04:58:03.845 INFO: [JOB 8473] sstable deleted 037071
        // 04:58:03.845 INFO: [JOB 8475] compacting: sstable created 037098
        // 04:58:03.849 INFO: [JOB 8475] compacted(default) L4 [037074] (989B) Score=5.86 + L5 [029331] (1.0KB) Score=0.67 -> L5 [037098] (989B), in 0.0s (0.0s total), output rate 609KB/s
        // 04:58:04.683 INFO: [JOB 8477] compacting(default) L4 [037094] (989B) Score=5.76 + L5 [029406] (1.0KB) Score=0.67; OverlappingRatio: Single 1.08, Multi 0.00
        // 04:58:04.684 INFO: [JOB 8475] sstable deleted 037074
        // 04:58:04.685 INFO: [JOB 8477] compacting: sstable created 037099
        // 04:58:04.688 INFO: [JOB 8477] compacted(default) L4 [037094] (989B) Score=5.76 + L5 [029406] (1.0KB) Score=0.67 -> L5 [037099] (989B), in 0.0s (0.0s total), output rate 650KB/s
        // 04:58:05.490 INFO: [JOB 8479] compacting(default) L3 [034661] (993B) Score=7.96 + L4 [028685] (1008B) Score=5.66; OverlappingRatio: Single 1.02, Multi 0.00
        // 04:58:05.491 INFO: [JOB 8477] sstable deleted 037094
        // 04:58:05.492 INFO: [JOB 8479] compacting: sstable created 037100
        // 04:58:05.492 INFO: [JOB 8479] compacting: sstable created 037101
        // 04:58:05.495 INFO: [JOB 8479] compacted(default) L3 [034661] (993B) Score=7.96 + L4 [028685] (1008B) Score=5.66 -> L4 [037100 037101] (1.9KB), in 0.0s (0.0s total), output rate 1.2MB/s
        // 04:58:06.296 INFO: [JOB 8481] compacting(default) L4 [037095] (989B) Score=5.71 + L5 [029408] (1.0KB) Score=0.67; OverlappingRatio: Single 1.08, Multi 0.00
        // 04:58:06.296 INFO: [JOB 8481] compacting: sstable created 037102
        // 04:58:06.299 INFO: [JOB 8481] compacted(default) L4 [037095] (989B) Score=5.71 + L5 [029408] (1.0KB) Score=0.67 -> L5 [037102] (989B), in 0.0s (0.0s total), output rate 3.1MB/s
        // 04:58:07.081 INFO: [JOB 8483] compacting(default) L4 [037101] (989B) Score=5.66 + L5 [029403] (1.0KB) Score=0.67; OverlappingRatio: Single 1.08, Multi 0.00
        // 04:58:07.082 INFO: [JOB 8481] sstable deleted 037095
        // 04:58:07.082 INFO: [JOB 8483] compacting: sstable created 037103
        // 04:58:07.086 INFO: [JOB 8483] compacted(default) L4 [037101] (989B) Score=5.66 + L5 [029403] (1.0KB) Score=0.67 -> L5 [037103] (989B), in 0.0s (0.0s total), output rate 607KB/s
        // 04:58:07.956 INFO: [JOB 8485] compacting(default) L3 [034663] (993B) Score=7.59 + L4 [028688] (1016B) Score=5.57; OverlappingRatio: Single 1.02, Multi 0.00
        // 04:58:07.957 INFO: [JOB 8483] sstable deleted 037101
        // 04:58:07.957 INFO: [JOB 8485] compacting: sstable created 037104
        // 04:58:07.957 INFO: [JOB 8485] compacting: sstable created 037105
        // 04:58:07.961 INFO: [JOB 8485] compacted(default) L3 [034663] (993B) Score=7.59 + L4 [028688] (1016B) Score=5.57 -> L4 [037104 037105] (1.9KB), in 0.0s (0.0s total), output rate 1.2MB/s
        // 04:58:08.803 INFO: [JOB 8487] compacting(default) L4 [036718] (989B) Score=5.61 + L5 [029365] (1.0KB) Score=0.67; OverlappingRatio: Single 1.08, Multi 0.00
        // 04:58:08.804 INFO: [JOB 8487] compacting: sstable created 037106
        // 04:58:08.807 INFO: [JOB 8487] compacted(default) L4 [036718] (989B) Score=5.61 + L5 [029365] (1.0KB) Score=0.67 -> L5 [037106] (989B), in 0.0s (0.0s total), output rate 3.1MB/s
        // 04:58:09.541 INFO: [JOB 8489] compacting(default) L4 [037105] (989B) Score=5.57 + L5 [029404] (1.0KB) Score=0.67; OverlappingRatio: Single 1.08, Multi 0.00
        // 04:58:09.542 INFO: [JOB 8489] compacting: sstable created 037107
        // 04:58:09.545 INFO: [JOB 8489] compacted(default) L4 [037105] (989B) Score=5.57 + L5 [029404] (1.0KB) Score=0.67 -> L5 [037107] (989B), in 0.0s (0.0s total), output rate 2.8MB/s
        // 04:58:10.362 INFO: [JOB 8491] compacting(default) L4 [037022] (1005B) Score=5.47 + L5 [029366] (1.1KB) Score=0.67; OverlappingRatio: Single 1.08, Multi 0.00
        // 04:58:10.364 INFO: [JOB 8489] sstable deleted 037105
        // 04:58:10.364 INFO: [JOB 8491] compacting: sstable created 037108
        // 04:58:10.367 INFO: [JOB 8491] compacted(default) L4 [037022] (1005B) Score=5.47 + L5 [029366] (1.1KB) Score=0.67 -> L5 [037108] (1005B), in 0.0s (0.0s total), output rate 626KB/s
        // 04:58:11.167 INFO: [JOB 8493] compacting(default) L3 [034684] (1001B) Score=7.27 + L4 [028923] (1.0KB) Score=5.37; OverlappingRatio: Single 1.03, Multi 0.00
        // 04:58:11.169 INFO: [JOB 8491] sstable deleted 037022
        // 04:58:11.169 INFO: [JOB 8493] compacting: sstable created 037109
        // 04:58:11.169 INFO: [JOB 8493] compacting: sstable created 037110
        // 04:58:11.172 INFO: [JOB 8493] compacted(default) L3 [034684] (1001B) Score=7.27 + L4 [028923] (1.0KB) Score=5.37 -> L4 [037109 037110] (1.9KB), in 0.0s (0.0s total), output rate 1.2MB/s
        // 04:58:11.923 INFO: [JOB 8495] compacting(default) L4 [033710] (827B) Score=5.41 + L5 [032289] (1.0KB) Score=0.67; OverlappingRatio: Single 1.24, Multi 0.00
        // 04:58:11.924 INFO: [JOB 8495] compacting: sstable created 037111
        // 04:58:11.927 INFO: [JOB 8495] compacted(default) L4 [033710] (827B) Score=5.41 + L5 [032289] (1.0KB) Score=0.67 -> L5 [037111] (1.0KB), in 0.0s (0.0s total), output rate 2.8MB/s
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMeta$' --run-dir /artifacts/meta/241017-045428.7741504490582/random-022 --try-to-reduce -v
        --- FAIL: TestMeta/execution/random-022 (223.36s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMeta$' ./internal/metamorphic -seed 1729140868774851973 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/17376495?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/17376495?buildTab=artifacts#meta) on refs/heads/master @ [dff8179025ee](https://github.com/cockroachdb/pebble/commits/dff8179025ee):


```
        // 04:58:00.518 INFO: [JOB 6947] compacting: sstable created 038511
        // 04:58:00.523 INFO: [JOB 6947] compacted(default) L5 [032803] (1.0KB) Score=96.79 + L6 [019969] (920B) Score=0.03 -> L6 [038511] (946B), in 0.0s (0.0s total), output rate 1.8MB/s
        // 04:58:01.573 INFO: [JOB 6949] compacting(default) L4 [031054] (1.0KB) Score=123.82 + L5 [] (0B) Score=96.30; OverlappingRatio: Single 0.00, Multi 0.96
        // 04:58:01.573 INFO: [JOB 6949] compacting: sstable created 038512
        // 04:58:01.577 INFO: [JOB 6949] compacted(default) L4 [031054] (1.0KB) Score=123.82 + L5 [] (0B) Score=96.30 -> L5 [038512] (1.0KB), in 0.0s (0.0s total), output rate 2.7MB/s
        // 04:58:02.592 INFO: [JOB 6951] compacting(default) L5 [034134] (1.0KB) Score=96.79 + L6 [038496] (946B) Score=0.03; OverlappingRatio: Single 0.91, Multi 0.00
        // 04:58:02.592 INFO: [JOB 6951] compacting: sstable created 038513
        // 04:58:02.592 INFO: [JOB 6951] compacting: sstable created 038514
        // 04:58:02.596 INFO: [JOB 6951] compacted(default) L5 [034134] (1.0KB) Score=96.79 + L6 [038496] (946B) Score=0.03 -> L6 [038513 038514] (1.8KB), in 0.0s (0.0s total), output rate 3.6MB/s
        // 04:58:03.723 INFO: [JOB 6952] compacting(default) L5 [034537] (1.0KB) Score=96.00 + L6 [038514] (920B) Score=0.03; OverlappingRatio: Single 0.88, Multi 0.00
        // 04:58:03.723 INFO: [JOB 6951] sstable deleted 038496
        // 04:58:03.723 INFO: [JOB 6952] compacting: sstable created 038515
        // 04:58:03.727 INFO: [JOB 6952] compacted(default) L5 [034537] (1.0KB) Score=96.00 + L6 [038514] (920B) Score=0.03 -> L6 [038515] (922B), in 0.0s (0.0s total), output rate 2.1MB/s
        // 04:58:04.789 INFO: [JOB 6953] compacting(default) L4 [031055] (1.0KB) Score=123.74 + L5 [] (0B) Score=95.51; OverlappingRatio: Single 0.00, Multi 0.96
        // 04:58:04.789 INFO: [JOB 6953] compacting: sstable created 038516
        // 04:58:04.792 INFO: [JOB 6952] sstable deleted 038514
        // 04:58:04.797 INFO: [JOB 6953] compacted(default) L4 [031055] (1.0KB) Score=123.74 + L5 [] (0B) Score=95.51 -> L5 [038516] (1.0KB), in 0.0s (0.0s total), output rate 638KB/s
        // 04:58:05.825 INFO: [JOB 6955] compacting(default) L5 [034046] (1.0KB) Score=96.01 + L6 [038487] (954B) Score=0.03; OverlappingRatio: Single 0.91, Multi 0.00
        // 04:58:05.825 INFO: [JOB 6955] compacting: sstable created 038517
        // 04:58:05.826 INFO: [JOB 6955] compacting: sstable created 038518
        // 04:58:05.830 INFO: [JOB 6955] compacted(default) L5 [034046] (1.0KB) Score=96.01 + L6 [038487] (954B) Score=0.03 -> L6 [038517 038518] (1.8KB), in 0.0s (0.0s total), output rate 1.5MB/s
        // 04:58:06.805 INFO: [JOB 6956] compacting(default) L4 [031056] (1.0KB) Score=122.37 + L5 [] (0B) Score=95.69; OverlappingRatio: Single 0.00, Multi 0.96
        // 04:58:06.805 INFO: [JOB 6956] compacting: sstable created 038519
        // 04:58:06.808 INFO: [JOB 6955] sstable deleted 038487
        // 04:58:06.811 INFO: [JOB 6956] compacted(default) L4 [031056] (1.0KB) Score=122.37 + L5 [] (0B) Score=95.69 -> L5 [038519] (1.0KB), in 0.0s (0.0s total), output rate 1.5MB/s
        // 04:58:07.939 INFO: [JOB 6958] compacting(default) L5 [038516] (1.0KB) Score=96.18 + L6 [021804] (1.0KB) Score=0.03; OverlappingRatio: Single 0.96, Multi 0.00
        // 04:58:07.939 INFO: [JOB 6958] compacting: sstable created 038520
        // 04:58:07.944 INFO: [JOB 6958] compacted(default) L5 [038516] (1.0KB) Score=96.18 + L6 [021804] (1.0KB) Score=0.03 -> L6 [038520] (959B), in 0.0s (0.0s total), output rate 1.6MB/s
        // 04:58:09.003 INFO: [JOB 6959] compacting(default) L5 [038519] (1.0KB) Score=95.70 + L6 [021828] (1.0KB) Score=0.03; OverlappingRatio: Single 0.96, Multi 0.00
        // 04:58:09.003 INFO: [JOB 6958] sstable deleted 038516
        // 04:58:09.003 INFO: [JOB 6959] compacting: sstable created 038521
        // 04:58:09.006 INFO: [JOB 6959] compacted(default) L5 [038519] (1.0KB) Score=95.70 + L6 [021828] (1.0KB) Score=0.03 -> L6 [038521] (1021B), in 0.0s (0.0s total), output rate 2.4MB/s
        // 04:58:10.065 INFO: [JOB 6960] compacting(default) L4 [031065] (1.1KB) Score=121.63 + L5 [] (0B) Score=94.74; OverlappingRatio: Single 0.00, Multi 0.95
        // 04:58:10.065 INFO: [JOB 6959] sstable deleted 038519
        // 04:58:10.065 INFO: [JOB 6960] compacting: sstable created 038522
        // 04:58:10.071 INFO: [JOB 6960] compacted(default) L4 [031065] (1.1KB) Score=121.63 + L5 [] (0B) Score=94.74 -> L5 [038522] (1.1KB), in 0.0s (0.0s total), output rate 2.4MB/s
        // 04:58:11.115 INFO: [JOB 6962] compacting(default) L5 [033819] (1.0KB) Score=95.23 + L6 [038482] (950B) Score=0.03; OverlappingRatio: Single 0.91, Multi 0.00
        // 04:58:11.115 INFO: [JOB 6962] compacting: sstable created 038523
        // 04:58:11.122 INFO: [JOB 6962] compacted(default) L5 [033819] (1.0KB) Score=95.23 + L6 [038482] (950B) Score=0.03 -> L6 [038523] (950B), in 0.0s (0.0s total), output rate 873KB/s
        // 04:58:12.159 INFO: [JOB 6963] compacting(default) L5 [038522] (1.1KB) Score=95.22 + L6 [023223] (1023B) Score=0.03; OverlappingRatio: Single 0.95, Multi 0.00
        // 04:58:12.159 INFO: [JOB 6962] sstable deleted 038482
        // 04:58:12.159 INFO: [JOB 6963] compacting: sstable created 038524
        // 04:58:12.163 INFO: [JOB 6963] compacted(default) L5 [038522] (1.1KB) Score=95.22 + L6 [023223] (1023B) Score=0.03 -> L6 [038524] (1.0KB), in 0.0s (0.0s total), output rate 2.4MB/s
        // 04:58:13.184 INFO: [JOB 6964] compacting(default) L4 [031066] (1.1KB) Score=120.87 + L5 [] (0B) Score=94.25; OverlappingRatio: Single 0.00, Multi 0.94
        // 04:58:13.184 INFO: [JOB 6963] sstable deleted 038522
        // 04:58:13.184 INFO: [JOB 6964] compacting: sstable created 038525
        // 04:58:13.188 INFO: [JOB 6964] compacted(default) L4 [031066] (1.1KB) Score=120.87 + L5 [] (0B) Score=94.25 -> L5 [038525] (1.1KB), in 0.0s (0.0s total), output rate 2.6MB/s
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMeta$' --run-dir /artifacts/meta/241019-045014.731172970075/random-004 --try-to-reduce -v
        --- FAIL: TestMeta/execution/random-004 (467.25s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMeta$' ./internal/metamorphic -seed 1729313414731594147 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/metamorphic/generator.go
+++ b/metamorphic/generator.go
@@ -1397,23 +1397,6 @@ func (g *generator) writerIngestExternalFiles() {
 				continue
 			}
 
-			// We can only use a synthetic suffix if we don't have multiple keys with
-			// the same prefix.
-			hasDuplicatePrefix := func() bool {
-				var prevPrefix []byte
-				for _, k := range g.keyManager.KeysForExternalIngest(objs[i]) {
-					prefix := g.prefix(k.key)
-					if g.cmp(prefix, prevPrefix) == 0 {
-						return true
-					}
-					prevPrefix = append(prevPrefix[:0], prefix...)
-				}
-				return false
-			}()
-			if hasDuplicatePrefix {
-				continue
-			}
-
 			// We can only use a synthetic suffix if we don't have overlapping range
 			// key sets (because they will become logically conflicting when we
 			// replace their suffixes with the synthetic one).

--- a/metamorphic/key_manager.go
+++ b/metamorphic/key_manager.go
@@ -307,6 +307,7 @@ func (k *keyManager) InRangeKeysForObj(o objID, lower, upper []byte) []keyMeta {
 // object (taking into consideration the bounds, synthetic suffix, etc).
 func (k *keyManager) KeysForExternalIngest(obj externalObjWithBounds) []keyMeta {
 	var res []keyMeta
+	var lastPrefix []byte
 	for _, km := range k.SortedKeysForObj(obj.externalObjID) {
 		// Apply prefix and suffix changes, then check the bounds.
 		if obj.syntheticPrefix.IsSet() {
@@ -316,6 +317,12 @@ func (k *keyManager) KeysForExternalIngest(obj externalObjWithBounds) []keyMeta 
 			n := k.comparer.Split(km.key)
 			km.key = append(km.key[:n:n], obj.syntheticSuffix...)
 		}
+		if lastPrefix != nil && k.comparer.Equal(lastPrefix, km.key[:k.comparer.Split(km.key)]) {
+			// We only keep the first of every unique prefix for external ingests.
+			// See the use of uniquePrefixes in newExternalObjOp.
+			continue
+		}
+		lastPrefix = append(lastPrefix[:0], km.key[:k.comparer.Split(km.key)]...)
 		if k.comparer.Compare(km.key, obj.bounds.Start) >= 0 && k.comparer.Compare(km.key, obj.bounds.End) < 0 {
 			res = append(res, km)
 		}

--- a/metamorphic/options.go
+++ b/metamorphic/options.go
@@ -651,9 +651,9 @@ func RandomOptions(
 	opts.Experimental.ValidateOnIngest = rng.IntN(2) != 0
 	opts.L0CompactionThreshold = 1 + rng.IntN(100)     // 1 - 100
 	opts.L0CompactionFileThreshold = 1 << rng.IntN(11) // 1 - 1024
-	opts.L0StopWritesThreshold = 1 + rng.IntN(100)     // 1 - 100
-	if opts.L0StopWritesThreshold < opts.L0CompactionThreshold {
-		opts.L0StopWritesThreshold = opts.L0CompactionThreshold
+	opts.L0StopWritesThreshold = 50 + rng.IntN(100)    // 50 - 150
+	if opts.L0StopWritesThreshold < 2*opts.L0CompactionThreshold {
+		opts.L0StopWritesThreshold = 2 * opts.L0CompactionThreshold
 	}
 	opts.LBaseMaxBytes = 1 << uint(rng.IntN(30)) // 1B - 1GB
 	maxConcurrentCompactions := rng.IntN(3) + 1  // 1-3

